ci-bench/src/main.rs

```diff
@@ -0,0 +1,309 @@
+use std::hint::black_box;
+use std::io::{self, Read, Write};
+
+use anyhow::Context;
+use clap::Parser;
+use rustls::{ClientConnection, ProtocolVersion, ServerConnection};
+
+use crate::bench_lib::{
+    validate_benchmarks, Benchmark, BenchmarkKind, BenchmarkParams, ResumptionKind,
+};
+use crate::cli::{Cli, Command, Side};
+use crate::util::config::{make_client_config, make_server_config, new_client, new_server};
+use crate::util::key::KeyType;
+use crate::util::transport::{
+    read_handshake_message, read_plaintext_to_end_bounded, send_handshake_message,
+    write_all_plaintext_bounded,
+};
+
+mod bench_lib;
+mod cli;
+mod util;
+
+/// The size in bytes of the plaintext sent in the transfer benchmark
+const TRANSFER_PLAINTEXT_SIZE: usize = 1024 * 1024;
+
+/// The amount of times a resumed handshake should be executed during benchmarking.
+///
+/// Handshakes with session resumption execute a very small amount of instructions (less than 200_000
+/// for some parameters), so a small difference in instructions accounts for a high difference in
+/// percentage (making the benchmark more sensitive to noise, because differences as low as 500
+/// instructions already raise a flag). Running the handshake multiple times gives additional weight
+/// to the instructions involved in the handshake, and less weight to noisy one-time setup code.
+///
+/// More specifically, great part of the noise in resumed handshakes comes from the usage of
+/// [`rustls::client::ClientSessionMemoryCache`] and [`rustls::server::ServerSessionMemoryCache`],
+/// which rely on a randomized `HashMap` under the hood (you can check for yourself by that
+/// `HashMap` by a `FxHashMap`, which brings the noise down to acceptable levels in a single run).
+const RESUMED_HANDSHAKE_RUNS: usize = 3;
+
+fn main() -> anyhow::Result<()> {
+    let benchmarks = get_benchmarks()?;
+
+    let cli = Cli::parse();
+    match cli.command {
+        Command::RunAll => {
+            let executable = std::env::args().next().unwrap();
```

Comment:
Is there some way to get the executable name as part of clap? The direct use of `std::env::args()` in the context of a clap parser feels bit odd.

Reply:
I agree it's odd, but clap doesn't seem to offer this through their API (the docs have no relevant references to the word `binary` or `executable` and googling did not reveal anything either)